packages/core/src/conflict-resolution.ts

```diff
@@ -242,14 +243,47 @@ export class ConflictResolution {
     private readonly handlers: HandlersMap
   ) {}
 
+  /**
+   * Helper function for applying a single commit to a StreamState.
+   * TODO: Most of this logic should be pushed down into the StreamHandler so it can be StreamType-specific.
+   * @param entry - the commit to apply
+   * @param state - the state to apply the commit to
+   * @param handler - the handler for the streamtype
+   * @private
+   */
+  private async applyLogEntryToState<T extends Stream>(
+    entry: CommitData, state: StreamState, handler: StreamHandler<T>): Promise<StreamState> {
+    const logEntry = await this.getCommitData(entry)
+    const commitMeta = { cid: logEntry.cid, timestamp: logEntry.timestamp }
+    if (StreamUtils.isAnchorCommit(logEntry.commit)) {
+      // It's an anchor commit
+      // TODO: Anchor validation should be done by the StreamHandler as part of applying the anchor commit
+      await verifyAnchorCommit(this.dispatcher, this.anchorValidator, logEntry.commit)
+      return await handler.applyCommit(logEntry.commit, commitMeta, this.context, state)
+    } else {
+      // It's a signed commit but may not be using DagJWS. If the JWS envelope is present, use that while applying the
+      // commit, otherwise use the commit itself (e.g. for CAIP-10 links).
+      const commitToApply = logEntry.envelope ? logEntry.envelope : logEntry.commit
+      const tmpState = await handler.applyCommit(commitToApply, commitMeta, this.context, state)
+      const isGenesis = !logEntry.commit.prev
+      const effectiveState = isGenesis ? tmpState : tmpState.next
+      // TODO: Schema validation should be done by the StreamHandler as part of applying the commit
+      await this.stateValidation.validate(effectiveState, effectiveState.content)
+      return tmpState // if validation is successful
+    }
+  }
+
   /**
    * Applies the log to the stream and updates the state.
+   * If an error is encountered while applying a commit, commit application stops and the state
+   * that was built thus far is returned, unless 'opts.throwOnApplyCommitError' is true.
    */
   private async applyLogToState<T extends Stream>(
     handler: StreamHandler<T>,
     unappliedCommits: CommitData[],
-    state?: StreamState,
-    breakOnAnchor?: boolean
+    state: StreamState | null,
+    breakOnAnchor: boolean,
```

Comment:
Can `breakOnAnchor` be folded into opts?

Reply:
It could be, but since it's only used in a single place I'd rather not as moving it to `InternalOpts` exposes it to a lot more layers of the stack unnecessarily.